main.rb

```diff
@@ -1,113 +1,222 @@
+# rubocop:disable Metrics/ModuleLength
+# rubocop:disable Metrics/MethodLength
+# rubocop:disable Metrics/CyclomaticComplexity
+# rubocop:disable Metrics/PerceivedComplexity
 module Enumerable
+  def make_array(input)
+    if input.class == Range
+      input.to_a
+    else
+      input
+    end
+  end
+
   def my_each
+    arr = make_array(self)
     if block_given?
-      for i in 0..self.length-1 do
-        yield(self[i])
+      i = 0
+      while i < arr.length
+        yield(arr[i])
+        i += 1
       end
-      self
     else
-      self.to_enum
+      arr.to_enum
     end
   end
 
-  def my_each_with_index
+  def my_each_2
+    arr = make_array(self)
     if block_given?
-      for i in 0..self.length-1 do
-        yield(self[i],i)
+      i = 1
+      while i < arr.length
+        yield(arr[i])
+        i += 1
       end
     else
-      self.to_enum
+      arr.to_enum
+    end
+  end
```

Comment:
- [ ] should be identical and returns the same thing as ruby's [`each_with_index`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-each_with_index). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each_with_index.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] my_each_with_index calls the given block once for each element in self
```ruby
      array = Array.new(100) { rand(0...9) } }
      my_each_output = ''
      block = proc { |num, idx| my_each_output += "Num: #{num}, idx: #{idx}\n" }
      array.each_with_index(&block)
      each_output = my_each_output.dup
      my_each_output = ''
      array.my_each_with_index(&block)
      my_each_output == each_output #true
```